### PR TITLE
mgr/orchestrator: Improve docstrings, add type hinting

### PR DESCRIPTION
**Pull request by sebastian-philipp**
1. Improve documentation
2. Improve IDE handling:
  screenshot_20181221_152539
  screenshot_20181221_154502

Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

- [x] Updates documentation if necessary
- [x] Does this work with Python 2?

### CODE REVIEW

File: src/pybind/mgr/orchestrator.py [hunk 1]
```diff
@@ -5,10 +5,16 @@
 Please see the ceph-mgr module developer's guide for more information.
 """
 
+from typing import TypeVar, Generic, List, Dict
```
Comment (tchaikov) [inline src/pybind/mgr/orchestrator.py]: `typing` was introduced by python 3.5. probably, we can catch `ImportError` and fallback to `object` if `typing` is not around? we can always use mypy in our own development environment if we want to perform some static analysis with python2. 

see http://pulpito.ceph.com/sage-2019-01-01_22:42:55-rados-wip-sage2-testing-2019-01-01-1414-distro-basic-smithi/

Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator.py]:
```
2019-01-01T23:55:48.808 INFO:tasks.ceph.mgr.x.smithi093.stderr:  File "/usr/lib64/ceph/mgr/ansible/__init__.py", line 5, in <module>
2019-01-01T23:55:48.808 INFO:tasks.ceph.mgr.x.smithi093.stderr:    from .module import Module
2019-01-01T23:55:48.808 INFO:tasks.ceph.mgr.x.smithi093.stderr:  File "/usr/lib64/ceph/mgr/ansible/module.py", line 10, in <module>
2019-01-01T23:55:48.808 INFO:tasks.ceph.mgr.x.smithi093.stderr:    import orchestrator
2019-01-01T23:55:48.809 INFO:tasks.ceph.mgr.x.smithi093.stderr:  File "/usr/lib64/ceph/mgr/orchestrator.py", line 8, in <module>
2019-01-01T23:55:48.809 INFO:tasks.ceph.mgr.x.smithi093.stderr:    from typing import TypeVar, Generic, List, Dict
2019-01-01T23:55:48.809 INFO:tasks.ceph.mgr.x.smithi093.stderr:ImportError: No module named typing
```

One problem is that catching `ImportError` breaks the IDE support. Guess I need to simplify this.

File: src/pybind/mgr/orchestrator.py [hunk 2]
```diff
@@ -5,10 +5,18 @@
...
+try:
+    from typing import TypeVar, Generic, List
+    T = TypeVar('T')
+    G = Generic[T]
+except ImportError:
+    T, G = object, object
 
-class _Completion(object):
+
+class _Completion(Generic[T]):
```
Comment (tserong) [inline src/pybind/mgr/orchestrator.py]: This probably needs to be `class _Completion(G)` for python 2 compatibility.

Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator.py]: fixed

**Comment by sebastian-philipp** [comment]
@tserong that's static type checking:

main.py:
```python
from typing import *

T = TypeVar('T')

class Completion(Generic[T]):
    def __init__(self):
        self.result: T = None

c = Completion[str]()
r = c.result
r.foobar
```
```
$ mypy main.py
main.py:11: error: "str" has no attribute "foobar"
```

This is similar to 

```cpp
#include <string>

template<typename T>
struct Completion
{
    T result = T();
};

int main()
{
    auto c = Completion<std::string>();
    auto r = c.result;
    r.foobar();
}
```
```
$ gcc -c main.cpp 
main.cpp: In function ‘int main()’:
main.cpp:13:7: error: ‘class std::__cxx11::basic_string<char>’ has no member named ‘foobar’
     r.foobar();
       ^~~~~~
```

**Comment by tserong** [comment]
> @tserong that's static type checking:

Got it, thanks.

File: src/pybind/mgr/orchestrator.py (hunk 1 above)
Comment (tchaikov) [reply src/pybind/mgr/orchestrator.py]: > One problem is that catching ImportError breaks the IDE support. Guess I need to simplify this.

@sebastian-philipp is it still true?

Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator.py]: nope. I fixed it

Comment (tchaikov) [reply src/pybind/mgr/orchestrator.py]: cool, so can we remove the "DNM" label?

Comment (sebastian-philipp) [reply src/pybind/mgr/orchestrator.py]: sure. sorry.

**Comment by sebastian-philipp** [comment]
jenkins render docs

**Comment by sebastian-philipp** [comment]
jenkins retest this please

**Comment by sebastian-philipp** [comment]
QA successful. Failure tracked in http://tracker.ceph.com/issues/37736

http://pulpito.ceph.com/swagner-2019-01-15_11:15:44-rados:mgr-wip-swagner-testing-distro-basic-smithi/

**Pull request merged**